core/dbt/rpc.py

```diff
@@ -66,3 +90,301 @@ def terminating(cls):
             cls.Error,
             cls.Result
         ]
+
+
+def sigterm_handler(signum, frame):
+    raise dbt.exceptions.RPCKilledException(signum)
+
+
+class RequestDispatcher(object):
+    """A special dispatcher that knows about requests."""
+    def __init__(self, http_request, json_rpc_request, manager):
+        self.http_request = http_request
+        self.json_rpc_request = json_rpc_request
+        self.manager = manager
+        self.task = None
+
+    def rpc_factory(self, task):
+        request_handler = RequestTaskHandler(task,
+                                             self.http_request,
+                                             self.json_rpc_request)
+
+        def rpc_func(**kwargs):
+            try:
+                self.manager.add_request(request_handler)
+                return request_handler.handle(kwargs)
+            finally:
+                self.manager.mark_done(request_handler)
+
+        return rpc_func
+
+    def __getitem__(self, key):
+        # the dispatcher's keys are method names and its values are functions
+        # that implement the RPC calls
+        func = self.manager.rpc_builtin(key)
+        if func is not None:
+            return func
+
+        task = self.manager.rpc_task(key)
+        return self.rpc_factory(task)
+
+
+class RequestTaskHandler(object):
+    def __init__(self, task, http_request, json_rpc_request):
+        self.task = task
+        self.http_request = http_request
+        self.json_rpc_request = json_rpc_request
+        self.queue = None
+        self.process = None
+        self.started = None
+        self.timeout = None
+        self.logs = []
+        self.task_id = uuid.uuid4()
+
+    @property
+    def request_source(self):
+        return self.http_request.remote_addr
+
+    @property
+    def request_id(self):
+        return self.json_rpc_request._id
+
+    @property
+    def method(self):
+        return self.task.METHOD_NAME
+
+    def _next_timeout(self):
+        if self.timeout is None:
+            return None
+        end = self.started + self.timeout
+        timeout = end - time.time()
+        if timeout < 0:
+            raise dbt.exceptions.RPCTimeoutException(self.timeout)
+        return timeout
+
+    def _wait_for_results(self):
+        """Wait for results off the queue. If there is a timeout set, and it is
+        exceeded, raise an RPCTimeoutException.
+        """
+        while True:
+            get_timeout = self._next_timeout()
+            try:
+                msgtype, value = self.queue.get(timeout=get_timeout)
+            except QueueEmpty:
+                raise dbt.exceptions.RPCTimeoutException(self.timeout)
```

Comment:
is this the right exception to raise here?

Reply:
Yes. if the queue has not received any messages in the calculated timeout value, we have exceeded the set timeout, and should raise a timeout exception. This is actually old code, it just used to live in `tasks/rpc_server.py`